### PR TITLE
DEV-520:ADD: Notifications API

### PR DESCRIPTION
**Pull request by Merisho**
- Generic subscription features added
- Major refactoring

### CODE REVIEW

File: internal/utils/utils.go [hunk 1]
```diff
@@ -14,3 +17,17 @@ func ISliceToStrSlice(s []interface{}) (stringSlice []string, err error) {
 
 	return stringSlice, nil
 }
+
+func StructToJSONMap(s interface{}) map[string]interface{} {
```
Comment (Pilatuz) [inline internal/utils/utils.go]: Have you seen the https://github.com/fatih/structs package or https://github.com/mitchellh/mapstructure

File: dh/client.go [hunk 2]
```diff
@@ -24,6 +25,56 @@ func (c *Client) Authenticate(token string) (result bool, err *Error) {
...
+	}
+
+	_, rawRes, err := c.request(data)
+
+	if err != nil {
+		return nil, "", err
+	}
+
+	type subsId struct {
+		Value int64 `json:"subscriptionId"`
```
[the remark this answers is not in the document]
Comment (Merisho) [reply dh/client.go]: Server responds with numeric value, so it won't be able to unmarshal JSON

**Comment by Merisho** [comment]
@Pilatuz Yes, I have tried to run tests with real DeviceHive, integration test suite located under **integrationtest** dir

File: internal/utils/utils.go (hunk 1 above)
Comment (Merisho) [reply internal/utils/utils.go]: Thank you, I have changed implementation

**Pull request merged**